lib/db.js

```diff
@@ -349,35 +379,45 @@ module.exports = (
 
   DB.prototype.devices = function (uid) {
     log.trace({ op: 'DB.devices', uid: uid })
+    const promises = [
+      this.pool.get('/account/' + uid + '/devices')
+    ]
 
-    return this.pool.get('/account/' + uid + '/devices')
-      .then(
-        function (body) {
-          return body.map(function (item) {
-            return {
-              id: item.id,
-              sessionToken: item.sessionTokenId,
-              lastAccessTime: marshallLastAccessTime(item.lastAccessTime, uid, item.email),
-              name: item.name,
-              type: item.type,
-              pushCallback: item.callbackURL,
-              pushPublicKey: item.callbackPublicKey,
-              pushAuthKey: item.callbackAuthKey,
-              uaBrowser: item.uaBrowser,
-              uaBrowserVersion: item.uaBrowserVersion,
-              uaOS: item.uaOS,
-              uaOSVersion: item.uaOSVersion,
-              uaDeviceType: item.uaDeviceType
-            }
-          })
-        },
-        function (err) {
-          if (isNotFoundError(err)) {
-            throw error.unknownAccount()
+    if (this.redis) {
+      promises.push(this.redis.getAsync(uid))
+    }
+    return P.all(promises)
+      .spread((devices, redisTokens) => {
+        const redisSessionTokens = redisTokens ? JSON.parse(redisTokens) : []
+
+        return devices.map(device => {
+          const token = redisSessionTokens.find(tok => tok.tokenId === device.sessionTokenId)
+          const hasToken = !! token
+          const lastAccessTime = hasToken ? token.lastAccessTime : device.lastAccessTime
+
+          return {
+            id: device.id,
+            sessionToken: device.sessionTokenId,
+            lastAccessTime: marshallLastAccessTime(lastAccessTime, uid, device.email),
+            name: device.name,
+            type: device.type,
+            pushCallback: device.callbackURL,
+            pushPublicKey: device.callbackPublicKey,
+            pushAuthKey: device.callbackAuthKey,
+            uaBrowser: hasToken ? token.uaBrowser : device.uaBrowser,
+            uaBrowserVersion: hasToken ? token.uaBrowserVersion : device.uaBrowserVersion,
+            uaOS: hasToken ? token.uaOS : device.uaOS,
+            uaOSVersion: hasToken ? token.uaOSVersion : device.uaOSVersion,
+            uaDeviceType: hasToken ? token.uaDeviceType : device.uaDeviceType
```

Comment:
Rather than explicitly checking `hasToken` 5 times, would this code be a little bit nicer if you leaned on the `Object.assign` trick you use in `DB.sessions`?